### PR TITLE
Add /lists/files/{filename} endpoints

### LINKED ISSUE #7180: Add /lists/files/{filename} endpoints
|Wazuh version|Component|
|---|---|
| 4.1.0 | Wazuh API |

Hello team,

as part of the epic issue https://github.com/wazuh/wazuh/issues/7148, we need to split the `manager/files` and `/cluster/{node_id}/files` endpoints.

This issue will cover lists files. For that, we need the following new endpoints:
- [x] `GET /lists/files/{filename}`
  - Add option to choose between raw or JSON.
  - **RBAC**: Action `lists:read` (it already exists). Resource `list:path`.
- [x] `PUT /lists/files/{filename}`
  - Raw format required.
  - **RBAC**: Action `lists:update`. Resource `list:path`.
- [x] `DELETE /lists/files/{filename}`
  - **RBAC**: Action `lists:delete`. Resource `list:path`.

These new endpoints require adaptations in the framework in order to have static paths and proper content validation. Remember to update the documentation for these new RBAC actions and resources.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [ ] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

Regards,
Víctor Fernández.

### PR DESCRIPTION
**Pull request by Selutario**
|Related issue|
|---|
| Closes #7180 |

## Description

Hey team!

As requested in #7180, this PR adds a new endpoint (`/lists/files/{filename}`) with three methods (GET, PUT and DELETE). 

Two new RBAC actions have been added too (`lists:update` and `lists:delete`), and the resource for the old `lists:read` action has been updated (`list:path:{path}` -> `list:file:{path}`), which required a refactor of the old CDB lists methods. 

- GET `/lists/files/{filename}`
  - RBAC actions: 
    - `lists:read`
  - RBAC resources:
    - `lists:file`
- PUT `/lists/files/{filename}`
  - RBAC actions: 
    - `lists:update`
    - `lists:delete` (if `overwrite` param is used)
  - RBAC resources:
    - `lists:file`
- DELETE `/lists/files/{filename}`
  - RBAC actions: 
    - `lists:delete`
  - RBAC resources:
    - `lists:file`

###  GET `/lists/files/{filename}`
Get the content of a CDB list. Only filename can be specified. It will be searched recursively inside `etc/lists` folder. 

> GET /lists/files/audit-keys

```JSON
{
  "data": {
    "affected_items": [
      {
        "audit-wazuh-w": "write",
        "audit-wazuh-r": "read",
        "audit-wazuh-a": "attribute",
        "audit-wazuh-x": "execute",
        "audit-wazuh-c": "command"
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "CDB list was returned",
  "error": 0
}
```

The output can be returned in raw, using the `raw` parameter too:

> GET /lists/files/audit-keys?raw=true

```audit-wazuh-w:write
audit-wazuh-r:read
audit-wazuh-a:attribute
audit-wazuh-x:execute
audit-wazuh-c:command
```

If the file is not found, error code 1802 is returned in JSON format even if `raw` is enabled:

> GET /lists/files/unknown?raw=true

```JSON
{
  "data": {
    "affected_items": [],
    "total_affected_items": 0,
    "total_failed_items": 1,
    "failed_items": [
      {
        "error": {
          "code": 1802,
          "message": "Lists file not found",
          "remediation": "Please, use `GET /lists/files` to find all available lists"
        },
        "id": [
          "unknown"
        ]
      }
    ]
  },
  "message": "No list was returned",
  "error": 1
}
```

###  PUT `/lists/files/{filename}`
Files can be uploaded or updated, only inside `etc/lists/`, and not any subdirectory:

> PUT /lists/files/new_file

```JSON
{
  "data": {
    "affected_items": [
      "etc/lists/new_file"
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "CDB list file uploaded successfully",
  "error": 0
}
```

However, it is checked that the name of the new file is not repeated, even inside any subdirectory of `etc/lists/`:

> PUT /lists/files/aws-eventnames

```JSON
{
  "data": {
    "affected_items": [],
    "total_affected_items": 0,
    "total_failed_items": 1,
    "failed_items": [
      {
        "error": {
          "code": 1805,
          "message": "File with the same name already exists in a subdirectory.",
          "remediation": "Please, make sure to use a name which is not repeated. "
        },
        "id": [
          "etc/lists/aws-eventnames"
        ]
      }
    ]
  },
  "message": "Could not upload CDB list file",
  "error": 1
}
```

The content of the file is validated before being uploaded or deleting the one to be replaced:

> PUT /lists/files/audit-keys?overwrite=true

```JSON
{
  "data": {
    "affected_items": [],
    "total_affected_items": 0,
    "total_failed_items": 1,
    "failed_items": [
      {
        "error": {
          "code": 1800,
          "message": "Bad format in CDB list {path}",
          "remediation": null
        },
        "id": [
          "etc/lists/audit-keys"
        ]
      }
    ]
  },
  "message": "Could not upload CDB list file",
  "error": 1
}
```

###  DELETE `/lists/files/{filename}`
Files can be deleted, only inside `etc/lists/`, and not any subdirectory. Both the list file and the `.cdb` (if exists) are deleted.

> DELETE /lists/files/new_file

```JSON
{
  "data": {
    "affected_items": [
      "etc/lists/new_file"
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "CDB list file was successfully deleted",
  "error": 0
}
```

## Unittests
New unit tests have been added, and most of the existing ones have been updated:
```
python3 -m pytest framework/wazuh/tests/test_cdb_list.py --disable-warnings 
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/framework
plugins: trio-0.6.0, asyncio-0.14.0
collected 54 items                                                                                                                                                     

framework/wazuh/tests/test_cdb_list.py ......................................................                                                                    [100%]

========================================================================== 54 passed in 0.23s ==========================================================================
```

```
python3 -m pytest framework/wazuh/core/tests/test_cdb_list.py --disable-warnings
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/framework
plugins: trio-0.6.0, asyncio-0.14.0
collected 39 items                                                                                                                                                     

framework/wazuh/core/tests/test_cdb_list.py .......................................                                                                              [100%]

========================================================================== 39 passed in 0.16s ==========================================================================
```

## Integration tests
API integration tests have also been updated and extended.

Regards,
Selu.

### CODE REVIEW

**Review by davidjiglesias — CHANGES_REQUESTED** [review]
Review all DAPIs from controllers. They must be local_master. The information is only updated live in the master node. 

Rename get_file, put_file and delete_file controllers to more specific list related names. For example_ get_list_file, put_list_file and delete_list_file

Please define specific error in API integration tests instead of error: !anyint, mainly for newly created or modified tests.

Review and remove all . from spec descriptions and API tests stage names.

File: api/test/integration/test_rbac_white_cdb_list_endpoints.tavern.yaml [hunk 1]
```diff
@@ -93,3 +93,134 @@ stages:
...
+        data:
+          failed_items:
+            - error:
+                code: 4000
+              id:
+                - 'etc/lists/security-eventchannel'
+          total_affected_items: 0
+          total_failed_items: 1
+
+  - name: Try to overwrite audit-keys list (allow)
```
Comment (davidjiglesias) [inline api/test/integration/test_rbac_white_cdb_list_endpoints.tavern.yaml:158]:
```suggestion
  - name: Overwrite audit-keys list (allow)
```

File: api/test/integration/test_rbac_white_cdb_list_endpoints.tavern.yaml [hunk 2]
```diff
@@ -93,3 +93,134 @@ stages:
...
+          affected_items: !anything
+          failed_items: [ ]
+          total_affected_items: 1
+          total_failed_items: 0
+
+---
+test_name: DELETE /lists/files/{filename}
+
+stages:
+  - name: Try to delete a CDB list (allow)
```
Comment (davidjiglesias) [inline api/test/integration/test_rbac_white_cdb_list_endpoints.tavern.yaml:200]:
```suggestion
  - name: Delete a CDB list (allow)
```

File: framework/wazuh/cdb_list.py [hunk 3]
```diff
@@ -3,44 +3,62 @@
...
 
     lists = list()
-    for rel_p in path:
-        if not any([relative_dirname is not None and os.path.dirname(rel_p) != relative_dirname,
-                    filename is not None and os.path.split(rel_p)[1] not in filename]):
-            lists.append({'items': get_list_from_file(rel_p),
-                          'relative_dirname': os.path.dirname(rel_p),
-                          'filename': os.path.split(rel_p)[1]})
+    for path in get_filenames_paths(filename):
+        if not any([dirname is not None and os.path.dirname(path) != dirname, not os.path.isfile(path)]):
```
Comment (davidjiglesias) [inline framework/wazuh/cdb_list.py:58]: Add comment, difficult piece of code to understand quickly

File: framework/wazuh/cdb_list.py [hunk 4]
```diff
@@ -52,33 +70,165 @@ def get_lists(path=None, offset=0, limit=common.database_limit, select=None, sor
...
+        # If file already exists and overwrite is False, raise exception.
+        if not overwrite and os.path.exists(path):
+            raise WazuhError(1905)
+        # If file with same name already exists in subdirectory.
+        elif get_filenames_paths([filename])[0] != path:
+            raise WazuhError(1805)
+        # Create backup and delete original CDB list.
+        elif overwrite and os.path.exists(path):
+            try:
+                backup_file = path + '.backup'
```
Comment (davidjiglesias) [inline framework/wazuh/cdb_list.py:142]:
```suggestion
                backup_file = f"{path}.backup"
```

Comment (davidjiglesias) [reply framework/wazuh/cdb_list.py:142]: Just for consistency with how it is done everywhere else

File: framework/wazuh/core/cdb_list.py [hunk 5]
```diff
@@ -128,48 +129,156 @@ def split_key_value_with_quotes(line, file_path='/CDB_LISTS_PATH'):
...
-        with open(file_path) as f:
-            for line in f.read().splitlines():
+        with open(path) as f:
+            output = f.read()
+
+        if raw:
+            result = output
+        else:
+            for line in output.splitlines():
                 if 'TEMPLATE' not in line:
```
Comment (davidjiglesias) [inline framework/wazuh/core/cdb_list.py:166]: We talked about checking TEMPLATE only as the key of the first line?

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

File: framework/wazuh/core/cdb_list.py (hunk 5 above)
Comment (Selutario) [reply framework/wazuh/core/cdb_list.py:166]: Replaced by a regex that will skip empty lines or line which start (not contain) `TEMPLATE:`

```PYTHON
def get_list_from_file(path, raw=False):
    """..."""
    # Match empty lines or lines which start with TEMPLATE:
    regex_template = re.compile(r'^TEMPLATE:|^\s*$')
    result = dict()

    try:
        with open(path) as f:
            output = f.read()

        if raw:
            result = output
        else:
            for line in output.splitlines():
                if not re.match(regex_template, line):
                    if '"' not in line:
                        # Check if key and value are not surrounded by double quotes
                        key, value = line.split(':')
                    else:
                        # Check if key and/or value are surrounded by double quotes
                        key, value = split_key_value_with_quotes(line, path)
                    result.update({key: value})
```

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**